### DISCUSSION TITLE
Feature: Support MergerFS and SnapRAID

### DISCUSSION THREAD

**Discussion by Protektor-Desura** [IceWhaleTech/CasaOS#220] [Ideas]
The issue is users are going to be adding maybe one drive at a time or couple of drives at a time over months to add more storage as they run out of space. It is unlikely for example that a home user will buy the total amount of storage that they will ever need. Instead it is far more likely that home users will want to buy more hard drives for more space as they run out.

The problem then becomes users won't want to have to reorganize everything all the time because they want more room for movies or books or whatever and have to keep all kinds of different directories or worry about how to use that last 10 gigs on one drive and that last 5 gigs on another drive.

The solution to this is to create a pool of all available storage drives in to one pool that the user doesn't have to worry about where free space is or how to add more space. This is where MergerFS comes in to play. It allows you to add one drive at a time and even add existing drives full of data to the pool.

https://github.com/trapexit/mergerfs

We already see support for MergerFS in a web interface in Open Media Vault. It is open source and possible to look at their web interface implementation and borrow ideas or even code from them to do this.

https://www.openmediavault.org/

The question then becomes how to we ensure that users don't loose data on their drives if they want redundancy? This can easily be solved by using and including SnapRAID. This is NOT a real time RAID solution but for home users who files do not often change it is a perfect solution because the pool can be expanded easily as users add one drive at a time over time.

https://github.com/amadvance/snapraid/
https://www.snapraid.it/

Again we already see SnapRAID supported in a web interface with Open Media Vault.

https://www.openmediavault.org/

The reason I do NOT recommend ZFS or BFRS or similar file systems or even software RAID is because you can NOT expand the pool one drive at a time with these solutions. You must have all the available drives at the time the pool is created and you can not easily expand the pool by a single drive once they are created.

Also if redundancy is NOT required by the home user then they can simply not use SnapRAID to provide redundancy for their files and simply pool all their drives in to one large data pool for storage.

**Comment by Protektor-Desura** [comment]
One of the reasons I am not a fan of Open Media Vault is I do not like the look of their web interface and they are based on Debian which tends to have older libraries and such which for creating a media server creates problems when you are trying to support GPU transcoding of videos.

**Reply by gpatkinson** [reply]
Agreed! Making it easy for a user to implement MergerFS/Snapraid with CasaOS would be fantastic! I've been looking around at other options including the awesome (but more CLI) https://perfectmediaserver.com and the MergerFS/Snapraid solution seems like a good one.

**Comment by tobias-rosenbaum** [comment]
I thought hard on whether to use CasaOS or OMV for my NAS build and settled on OMV because I wanted to have the MergerFS and SnapRAID combination. If CasaOS featured this, I would have chosen CasaOS over OMV.
